misk-cron/src/main/kotlin/misk/cron/CronManager.kt

```diff
@@ -114,6 +138,45 @@ class CronManager @Inject constructor() {
     }
   }
 
+  private fun tryAcquireLeasePerCronAndRunCrons(lastRun: Instant) {
+    val now = clock.instant()
+    val previousTime = ZonedDateTime.ofInstant(lastRun, zoneId)
+    logger.info {
+      "Last execution was at $previousTime, now=${ZonedDateTime.ofInstant(now, zoneId)}"
+    }
+    removeCompletedCrons()
+    cronEntries.values.forEach { cronEntry ->
+      val taskLease = buildTaskLease(cronEntry.runnable::class)
+      val holdsTaskLease = if (!taskLease.checkHeld()) {
+        taskLease.acquire()
+      } else {
+        true
+      }
+      if (!holdsTaskLease) {
+        return@forEach
+      }
+
+      val nextExecutionTime = cronEntry.executionTime.nextExecution(previousTime).orElseThrow()
+        .withSecond(0)
+        .withNano(0)
+
+
+      if (nextExecutionTime.toInstant() <= now) {
+        logger.info {
+          "CronJob ${cronEntry.name} was ready at $nextExecutionTime"
+        }
+        runCron(cronEntry)
+        taskLease.release()
+      }
+    }
+  }
+
+  internal fun buildTaskLease(klass: KClass<out Runnable>): Lease {
+    val sanitizedClassName = klass.qualifiedName!!.replace(".", "-")
```

Comment:
There are some gotchas here with inner/anonymous classes. You might end up with some very strange looking lease names. Is this replacement just for readability or are there actual characters/sequences you intend to avoid in the lease name string?

Reply:
We mainly weren't sure if the partition key had forbidden characters, and since the dynamo lease implementation lives in `Skim` it wasn't straightforward to write an integration test. Looking more thoroughly at the [dynamo docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html), it doesn't seem like that's the case though.

Reply:
Removed the sanitization, `.` are clearly fine since the original cluster wide lease uses them